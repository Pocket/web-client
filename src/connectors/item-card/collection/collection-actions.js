import React from 'react'
import { SaveToPocket } from 'components/item-actions/save-to-pocket'
import { useSelector, useDispatch } from 'react-redux'
import { saveCollectionPage } from 'containers/collections/collections.state'

import { itemActionStyle } from 'components/item-actions/base'

import { trackItemSave } from 'connectors/snowplow/snowplow.state'
import { trackItemOpen } from 'connectors/snowplow/snowplow.state'

export function ActionsCollection({ id, position }) {
  const dispatch = useDispatch()
  const isAuthenticated = useSelector((state) => state.user.auth)
  const item = useSelector((state) => state.collectionsBySlug[id])

  if (!item) return null

  const { url, pageSaveStatus } = item

  // Prep save action
  const onSave = () => {
    dispatch(saveCollectionPage(id))
    dispatch(trackItemSave(position, { url }, 'collection.save'))
  }

  // Open action
  const onOpen = () => dispatch(trackItemOpen(position, { url }, 'collection.open', url))

  return item ? (
    <div className={`${itemActionStyle} actions`}>
      <SaveToPocket
        allowRead={false}
        url={url}
        onOpen={onOpen}
        openExternal={false}
        saveAction={onSave}
        isAuthenticated={isAuthenticated}
        saveStatus={pageSaveStatus}
        id={id}
      />
    </div>
  ) : null
}
