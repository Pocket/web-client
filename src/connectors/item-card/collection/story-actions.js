import React from 'react'
import { SaveToPocket } from 'components/item-actions/save-to-pocket'
import { useSelector, useDispatch } from 'react-redux'
import { saveItem } from 'connectors/items-by-id/collection/items.state'

import { itemActionStyle } from 'components/item-actions/base'

import { trackItemSave } from 'connectors/snowplow/snowplow.state'
import { trackItemOpen } from 'connectors/snowplow/snowplow.state'

export function ActionsCollection({ id, position }) {
  const dispatch = useDispatch()
  const isAuthenticated = useSelector((state) => state.user.auth)
  const item = useSelector((state) => state.collectionItemsById[id])

  if (!item) return null

  const { url, save_status = 'unsaved' } = item

  // Prep save action
  const onSave = () => {
    dispatch(saveItem(id, url))
    dispatch(trackItemSave(position, { url }, 'collection.save'))
  }

  // Open action
  const onOpen = () => dispatch(trackItemOpen(position, { url }, 'collection.open', url))

  return item ? (
    <div className={`${itemActionStyle} actions`}>
      <SaveToPocket
        allowRead={false}
        url={url}
        onOpen={onOpen}
        openExternal={false}
        saveAction={onSave}
        isAuthenticated={isAuthenticated}
        saveStatus={save_status}
        id={id}
      />
    </div>
  ) : null
}
