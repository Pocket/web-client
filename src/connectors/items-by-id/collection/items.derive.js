import { BASE_URL } from 'common/constants'

export function deriveCollectionItems(response) {
  /**
 * @title {string} The most appropriate title to show
 * @thumbnail {url} The most appropriate image to show as a thumbnail
 * @publisher {string} The best text to display as the publisher of this item
 * @excerpt {string} The most appropriate excerpt to show
 * @open_url {url} The url that should be saved or opened
 * @share_url {url} The url that should be shared if the user shares this item
 * @save_status {string} A string value (unsaved, saving, saved)
 */
    return response.map((collectionItem) => {
    return {
      slug: collectionItem.slug,
      title: collectionItem.title,
      thumbnail: collectionItem.thumbnail,
      publisher: 'Pocket',
      excerpt: collectionItem.excerpt,
      save_url: collectionUrl(collectionItem),
      open_url: collectionUrl(collectionItem),
      syndicated: false,
      original_url: collectionUrl(collectionItem),
      openExternal: false,
      save_status: 'unsaved',
      authors: collectionItem.authors
    }
  })
}

function collectionUrl(collectionItem) {
  return `/collections/${collectionItem.slug}`
}

