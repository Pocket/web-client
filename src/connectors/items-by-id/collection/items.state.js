import { put, takeEvery } from 'redux-saga/effects'

import { arrayToObject } from 'common/utilities'
import { getCollections as apiFetchCollections } from 'common/api/collections'
import { deriveCollectionItems } from 'connectors/items-by-id/collection/items.derive'

import { saveItem as saveItemAPI } from 'common/api/saveItem'

import { COLLECTION_ITEMS_REQUEST } from 'actions'
import { COLLECTION_ITEMS_REQUEST_SUCCESS } from 'actions'
import { COLLECTION_ITEMS_REQUEST_FAILURE } from 'actions'
import { COLLECTION_ITEMS_HYDRATE } from 'actions'
import { COLLECTION_ITEMS_SAVE_REQUEST } from 'actions'
import { COLLECTION_ITEMS_SAVE_SUCCESS } from 'actions'
import { COLLECTION_ITEMS_SAVE_FAILURE } from 'actions'
import { COLLECTION_ITEMS_NO_IMAGE } from 'actions'

import { HYDRATE } from 'actions'

/** ACTIONS
 --------------------------------------------------------------- */
export const getCollectionsItems = () => ({ type: COLLECTION_ITEMS_REQUEST })
export const hydrateItems = (hydrated) => ({ type: COLLECTION_ITEMS_HYDRATE, hydrated }) //prettier-ignore
export const saveItem = (id, url) => ({type: COLLECTION_ITEMS_SAVE_REQUEST, id, url}) //prettier-ignore
export const setNoImage = (id) => ({ type: COLLECTION_ITEMS_NO_IMAGE, id })

/** REDUCERS
 --------------------------------------------------------------- */
const initialState = {}

export const collectionItemsReducers = (state = initialState, action) => {
  switch (action.type) {
    case COLLECTION_ITEMS_HYDRATE: {
      const { hydrated } = action
      return { ...state, ...hydrated }
    }

    case COLLECTION_ITEMS_SAVE_REQUEST: {
      const { id } = action
      return updateSaveStatus(state, id, 'saving')
    }

    case COLLECTION_ITEMS_SAVE_SUCCESS: {
      const { id } = action
      return updateSaveStatus(state, id, 'saved')
    }

    case COLLECTION_ITEMS_SAVE_FAILURE: {
      const { id } = action
      return updateSaveStatus(state, id, 'unsaved')
    }

    case COLLECTION_ITEMS_NO_IMAGE: {
      const { id } = action
      const item = state[id]
      const itemDraft = { ...item, noImage: true }
      return { ...state, [id]: itemDraft }
    }

    case COLLECTION_ITEMS_REQUEST_SUCCESS: {
      const { collectionItemsById } = action
      return { ...state, ...collectionItemsById }
    }

    // SPECIAL HYDRATE:  This is sent from the next-redux wrapper and
    // it represents the state used to build the page on the server.
    case HYDRATE:
      const { collectionItemsById } = action.payload
      return { ...state, ...collectionItemsById }

    default:
      return state
  }
}

/** UPDATE SAVE STATUS
 * Helper function to update save status for a specific item based on id
 * @param {object} state Redux state object
 * @param {string} id Item id to operate on
 * @param {string} save_status Value to update save status to
 */
export function updateSaveStatus(state, id, save_status) {
  const updatedItem = { ...state[id], save_status }
  return { ...state, [id]: updatedItem }
}

/** SAGAS :: WATCHERS
 --------------------------------------------------------------- */
export const collectionItemsSagas = [
  takeEvery(COLLECTION_ITEMS_SAVE_REQUEST, itemsSaveRequest),
  takeEvery(COLLECTION_ITEMS_REQUEST, collectionItemsRequest)
]

/** SAGA :: RESPONDERS
 --------------------------------------------------------------- */

function* itemsSaveRequest(action) {
  try {
    const { url, id } = action

    const response = yield saveItemAPI(url)
    if (response?.status !== 1) throw new Error('Unable to save')

    yield put({ type: COLLECTION_ITEMS_SAVE_SUCCESS, id })
  } catch (error) {
    yield put({ type: COLLECTION_ITEMS_SAVE_FAILURE, error })
  }
}

function* collectionItemsRequest(action) {
  try {
    const collectionItemsById = yield fetchCollections()
    yield put({ type: COLLECTION_ITEMS_REQUEST_SUCCESS, collectionItemsById })
  } catch (error) {
    yield put({ type: COLLECTION_ITEMS_REQUEST_FAILURE, error })
  }
}

export async function fetchCollections() {
  try {
    const response = await apiFetchCollections()

    if (!response) return { error: 'No data found' }

    const derivedCollections = await deriveCollectionItems(response)
    const collectionItemsById = arrayToObject(derivedCollections, 'slug')

    return collectionItemsById
  } catch (error) {
    console.log('items.state.fetchCollections', error)
  }
}
