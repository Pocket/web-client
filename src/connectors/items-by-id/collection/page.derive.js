export const deriveCollectionPage = (response) => {
  return response.map((collection) => {
    const slug = collection.slug
    const url = `/collections/${slug}`
    const firstImage = collection.stories[0].thumbnail
    const authorImage = collection.authors[0].imageUrl
    const heroImage = collection.thumbnail
    return {
      ...collection,
      heroImage: heroImage,
      thumbnail: authorImage || firstImage,
      url,
      item_id: slug
    }
  })
}
