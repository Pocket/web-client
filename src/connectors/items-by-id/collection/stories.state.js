import { put, takeEvery } from 'redux-saga/effects'
import { saveItem as saveItemAPI } from 'common/api/saveItem'
import { deriveStory } from 'common/api/derivers/item'
import { arrayToObject } from 'common/utilities'

import { COLLECTION_STORIES_REQUEST } from 'actions'
import { COLLECTION_STORIES_HYDRATE } from 'actions'
import { COLLECTION_STORIES_HYDRATE_SUCCESS } from 'actions'
import { COLLECTION_STORIES_HYDRATE_FAILURE } from 'actions'
import { COLLECTION_STORIES_SAVE_REQUEST } from 'actions'
import { COLLECTION_STORIES_SAVE_SUCCESS } from 'actions'
import { COLLECTION_STORIES_SAVE_FAILURE } from 'actions'
import { COLLECTION_STORIES_NO_IMAGE } from 'actions'
import { HYDRATE } from 'actions'

/** ACTIONS
 --------------------------------------------------------------- */
export const getCollectionsStories = () => ({ type: COLLECTION_STORIES_REQUEST })
export const hydrateStories = (stories) => ({ type: COLLECTION_STORIES_HYDRATE, stories }) //prettier-ignore
export const saveStory = (id, url) => ({type: COLLECTION_STORIES_SAVE_REQUEST, id, url}) //prettier-ignore
export const setNoImage = (id) => ({ type: COLLECTION_STORIES_NO_IMAGE, id })

/** REDUCERS
 --------------------------------------------------------------- */
const initialState = {}

export const collectionStoriesReducers = (state = initialState, action) => {
  switch (action.type) {
    case COLLECTION_STORIES_HYDRATE_SUCCESS: {
      const { storiesById } = action
      return { ...state, ...storiesById }
    }

    case COLLECTION_STORIES_SAVE_REQUEST: {
      const { id } = action
      return updateSaveStatus(state, id, 'saving')
    }

    case COLLECTION_STORIES_SAVE_SUCCESS: {
      const { id } = action
      return updateSaveStatus(state, id, 'saved')
    }

    case COLLECTION_STORIES_SAVE_FAILURE: {
      const { id } = action
      return updateSaveStatus(state, id, 'unsaved')
    }

    case COLLECTION_STORIES_NO_IMAGE: {
      const { id } = action
      const item = state[id]
      const itemDraft = { ...item, noImage: true }
      return { ...state, [id]: itemDraft }
    }

    // SPECIAL HYDRATE:  This is sent from the next-redux wrapper and
    // it represents the state used to build the page on the server.
    case HYDRATE: {
      const { collectionStoriesById } = action.payload
      return { ...state, ...collectionStoriesById }
    }

    default:
      return state
  }
}

/** UPDATE SAVE STATUS
 * Helper function to update save status for a specific item based on id
 * @param {object} state Redux state object
 * @param {string} id Item id to operate on
 * @param {string} save_status Value to update save status to
 */
export function updateSaveStatus(state, id, saveStatus) {
  const updatedItem = { ...state[id], saveStatus }
  return { ...state, [id]: updatedItem }
}

/** SAGAS :: WATCHERS
 --------------------------------------------------------------- */
export const collectionStoriesSagas = [
  takeEvery(COLLECTION_STORIES_SAVE_REQUEST, storiesSaveRequest),
  takeEvery(COLLECTION_STORIES_HYDRATE, storiesDerive)
]

/** SAGA :: RESPONDERS
 --------------------------------------------------------------- */

function* storiesSaveRequest(action) {
  try {
    const { url, id } = action

    const response = yield saveItemAPI(url)
    if (response?.status !== 1) throw new Error('Unable to save')

    yield put({ type: COLLECTION_STORIES_SAVE_SUCCESS, id })
  } catch (error) {
    yield put({ type: COLLECTION_STORIES_SAVE_FAILURE, error })
  }
}

function* storiesDerive(action) {
  try {
    const { stories } = action
    const derivedStories = stories.map((story) => deriveStory(story))
    const storiesById = arrayToObject(derivedStories, 'itemId')

    yield put({ type: COLLECTION_STORIES_HYDRATE_SUCCESS, storiesById })
  } catch (error) {
    yield put({ type: COLLECTION_STORIES_HYDRATE_FAILURE, error })
  }
}
