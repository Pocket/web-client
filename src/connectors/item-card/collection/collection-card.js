import { Card } from 'components/item-card/card'
import { useSelector, useDispatch } from 'react-redux'
import { ActionsCollection } from './collection-actions'
import { setNoImage } from 'connectors/items-by-id/discover/items.state'
import { trackItemImpression } from 'connectors/snowplow/snowplow.state'
import { trackItemOpen } from 'connectors/snowplow/snowplow.state'

/**
 * Article Card
 * Creates a connected `Card` with the appropriate data and actions
 * @param {object} {id, position} item_id for data and position for analytics
 */
export function ItemCard({ id, cardShape, className, showExcerpt = false, position }) {
  const dispatch = useDispatch()

  // Get data from state
  const item = useSelector((state) => state.collectionsBySlug[id])
  const { url } = item
  const impressionFired = useSelector((state) => state.analytics.impressions.includes(id))
  const onImageFail = () => dispatch(setNoImage(id))

  /**
   * ITEM TRACKING
   * ----------------------------------------------------------------
   */
  const onImpression = () => dispatch(trackItemImpression(position, item, 'collection.impression'))
  const onItemInView = (inView) => (!impressionFired && inView ? onImpression() : null)
  const onOpen = () => dispatch(trackItemOpen(position, item, 'collection.open'))

  return (
    <Card
      item={item}
      onImageFail={onImageFail}
      position={position}
      className={className}
      cardShape={cardShape}
      showExcerpt={showExcerpt}
      openUrl={url}
      useMarkdown={true}
      // Tracking
      onItemInView={onItemInView}
      onOpen={onOpen}
      // Actions
      actionId={item.slug}
      ActionMenu={ActionsCollection}
    />
  )
}
