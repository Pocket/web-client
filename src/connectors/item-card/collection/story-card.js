import { Card } from 'components/item-card/card'
import { useSelector, useDispatch } from 'react-redux'
import { ActionsCollection } from './story-actions'
import { setNoImage } from 'connectors/items-by-id/discover/items.state'
import { trackItemImpression } from 'connectors/snowplow/snowplow.state'
import { trackItemOpen } from 'connectors/snowplow/snowplow.state'

/**
 * Article Card
 * Creates a connected `Card` with the appropriate data and actions
 * @param {object} {id, position} item_id for data and position for analytics
 */
export function ItemCard({ id, cardShape, className, showExcerpt = false, position }) {
  const dispatch = useDispatch()
  // Get data from state
  const item = useSelector((state) => state.collectionStoriesById[id])
  const impressionFired = useSelector((state) => state.analytics.impressions.includes(id))

  if (!item) return null

  const { open_url: url } = item
  const onImageFail = () => dispatch(setNoImage(id))

  /**
   * ITEM TRACKING
   * ----------------------------------------------------------------
   */
  const onImpression = () => dispatch(trackItemImpression(position, { url }, 'collection.impression'))
  const onItemInView = (inView) => (!impressionFired && inView ? onImpression() : null)
  const onOpen = () => dispatch(trackItemOpen(position, { url }, 'collection.open', url))

  return (
    <Card
      item={item}
      onImageFail={onImageFail}
      position={position}
      className={className}
      cardShape={cardShape}
      showExcerpt={showExcerpt}
      openUrl={url}
      useMarkdown={true}
      // Tracking
      onItemInView={onItemInView}
      onOpen={onOpen}
      ActionMenu={ActionsCollection}
    />
  )
}
