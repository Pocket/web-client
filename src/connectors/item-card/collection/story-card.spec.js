import { wrappedRender } from 'test-utils'
import '@testing-library/jest-dom/extend-expect'
import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils'

import { storyFromClientApi } from 'common/api/derivers/collection/story.spec'
import { ItemCard as StoryCard } from 'connectors/item-card/collection/story-card'
import { deriveStory } from 'common/api/derivers/item'

describe('ItemCard', () => {
  //Legacy Derivers
  const derivedStory = deriveStory(storyFromClientApi)

  const initialState = {
    user: { auth: true },
    analytics: {
      impressions: []
    },
    collectionStoriesById: {
      [derivedStory.itemId]: derivedStory
    }
  }

  beforeAll(() => {
    mockAllIsIntersecting()
  })

  it('renders a story', () => {
    const { getByCy } = wrappedRender(<StoryCard id={derivedStory.itemId} position={3} />, { initialState }) //prettier-ignore
    expect(getByCy('article-card-', { exact: false })).toMatchSnapshot()
  })
})
