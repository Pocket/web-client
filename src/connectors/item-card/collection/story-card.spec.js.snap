// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard renders a story 1`] = `
<article
  class="c18o9ext grid noExcerpt"
  data-cy="article-card-1731163180"
>
  <div
    class="selectedBack"
  />
  <div
    class="cardWrap"
  >
    <div
      class="c97c5c media"
    >
      <a
        data-cy="image-link"
        href="https://www.nytimes.com/2017/05/06/business/inside-vws-campaign-of-trickery.html?utm_source=pocket_collection_story"
        rel="noopener noreferrer"
        tabindex="-1"
        target="_blank"
      >
        <img
          alt=""
          src="https://pocket-image-cache.com/600x400/filters:format(jpg):extract_focal()/https%3A%2F%2Fs3.amazonaws.com%2Fpocket-collectionapi-prod-images%2F914953af-15fd-46b9-ba3c-eed3ed278f9a.jpeg"
          srcset=" https://pocket-image-cache.com/1200x800/filters:format(jpg):extract_focal()/https%3A%2F%2Fs3.amazonaws.com%2Fpocket-collectionapi-prod-images%2F914953af-15fd-46b9-ba3c-eed3ed278f9a.jpeg 2x "
          style="--fallbackBackground: #FCB64380; --fallbackColor: #FCB643; --fallbackLetter: 'I';"
        />
        <span
          class="view-original"
          data-cy="view-original"
        >
          View Original 
          <span
            class="i1qqph0t icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M6 4a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2v-3a1 1 0 1 1 2 0v3a4 4 0 0 1-4 4H6a4 4 0 0 1-4-4V6a4 4 0 0 1 4-4h3a1 1 0 0 1 0 2H6Z"
                fill-rule="evenodd"
              />
              <path
                clip-rule="evenodd"
                d="M13 3a1 1 0 0 1 1-1h7a1 1 0 0 1 1 1v7a1 1 0 1 1-2 0V5.414l-7.293 7.293a1 1 0 0 1-1.414-1.414L18.586 4H14a1 1 0 0 1-1-1Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </span>
      </a>
    </div>
    <div
      class="content"
    >
      <h2
        class="title openExternal"
      >
        <a
          data-cy="title-link"
          href="https://www.nytimes.com/2017/05/06/business/inside-vws-campaign-of-trickery.html?utm_source=pocket_collection_story"
          rel="noopener noreferrer"
          tabindex="0"
          target="_blank"
        >
          Inside VWâ€™s Campaign of Trickery
        </a>
        <span
          class="i1qqph0t icon mobile-view-original"
          data-cy="view-original-icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M6 4a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2v-3a1 1 0 1 1 2 0v3a4 4 0 0 1-4 4H6a4 4 0 0 1-4-4V6a4 4 0 0 1 4-4h3a1 1 0 0 1 0 2H6Z"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M13 3a1 1 0 0 1 1-1h7a1 1 0 0 1 1 1v7a1 1 0 1 1-2 0V5.414l-7.293 7.293a1 1 0 0 1-1.414-1.414L18.586 4H14a1 1 0 0 1-1-1Z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </h2>
      <cite
        class="details"
      >
        <div
          class="authors"
        >
          <span>
            Jack Ewing
          </span>
        </div>
        <a
          class="publisher"
          data-cy="publisher-link"
          href="https://www.nytimes.com/2017/05/06/business/inside-vws-campaign-of-trickery.html?utm_source=pocket_collection_story"
          rel="noopener noreferrer"
          tabindex="0"
          target="_blank"
        >
          New York Times
        </a>
      </cite>
    </div>
  </div>
  <footer
    class="footer"
  >
    <div
      class="i18uycg6 actions"
    >
      <button
        class="b5bt6fr sb4ns5w unsaved card-actions"
        data-cy="article-save-btn-1731163180"
      >
        <span
          class="i1qqph0t icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M2 5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2H2Zm2 0H2v6c0 5.523 4.477 10 10 10s10-4.477 10-10V5h-2v6a8 8 0 1 1-16 0V5Z"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M7.293 9.293a1 1 0 0 1 1.414 0L12 12.586l3.293-3.293a1 1 0 1 1 1.414 1.414l-4 4a1 1 0 0 1-1.414 0l-4-4a1 1 0 0 1 0-1.414Z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <div
          class="actionCopy"
        >
          item-action:save-unsaved
        </div>
      </button>
    </div>
  </footer>
</article>
`;
